src/s_audio_pulse.c

```diff
@@ -0,0 +1,171 @@
+    int err = 0;
+
+    if(channels <= 0)
+        return client;
+    if(channels > 255)
+        channels = 255;
+    if(blocksize < DEFDACBLKSIZE)
+        blocksize = DEFDACBLKSIZE;
+
+    client = getbytes(sizeof(t_pd_pulseclient));
+    if(!client)
+        return 0;
+
+    client->spec.format = PA_SAMPLE_FLOAT32;
+    client->spec.rate = rate;
+    client->spec.channels = channels;
+    client->bufsize = sizeof(*client->buffer) * channels * DEFDACBLKSIZE;
+    client->buffer = getbytes(client->bufsize);
+
+    client->bufattr.maxlength = -1;
+    client->bufattr.tlength = sizeof(float) * channels * blocksize;
+    client->bufattr.prebuf = -1;
+    client->bufattr.minreq = -1;
+    client->bufattr.fragsize = sizeof(float) * channels * DEFDACBLKSIZE;
+
+    client->client = pa_simple_new(
+        NULL,               // Use the default server.
+        client_name,        // Our application's name.
+        dir,
+        NULL,               // Use the default device.
+        stream_name,        // Description of our stream.
+        &client->spec,      // Our sample format.
+        NULL,               // Use default channel map
+        &client->bufattr,   // buffer attributes
+        &err);
+
+    if (!client->client)
+    {
+        error("Failed to initialize PulseAudio '%s': %s", stream_name, pa_strerror(err));
+    }
+
+    if (!client->buffer || !client->client) {
+        client = free_client(client);
+    }
+    if(client && sys_verbose)
+    {
+        pa_usec_t latency;
+        if ((latency = pa_simple_get_latency(client->client, &err)) == (pa_usec_t) -1) {
+            error("pa_simple_get_latency() failed: %s", pa_strerror(err));
+        } else
+            post("PulseAudio latency '%s':\t%0.0f ms (%d bytes)", stream_name, (float)latency/1000., client->bufattr.tlength);
+    }
+    return client;
+}
+
+int
+pulse_open_audio(int inchans, int outchans, int rate, int blocksize)
+{
+    pulse_sink = make_client(PA_STREAM_PLAYBACK, outchans, rate, blocksize);
+    pulse_source = make_client(PA_STREAM_RECORD, inchans, rate, blocksize);
+
+    return !(pulse_sink || pulse_source);
+}
+
+void pulse_close_audio(void)
+{
+    pulse_sink = free_client(pulse_sink);
+    pulse_source = free_client(pulse_source);
+}
+
+int pulse_send_dacs(void)
+{
+
+    if (!STUFF->st_inchannels && !STUFF->st_outchannels) return (SENDDACS_NO);
+
+    if (STUFF->st_inchannels && pulse_source) {
+        int err = 0;
+        t_sample * fp = STUFF->st_soundin;
+        if(pa_simple_read(pulse_source->client, pulse_source->buffer, pulse_source->bufsize, &err) < 0)
+            error("Pulse Audio: %s", pa_strerror(err));
+        else {
+            const int channels = STUFF->st_inchannels;
+            int chan, frame;
+            float*buf = pulse_source->buffer;
+            for(chan = 0; chan<channels; chan++)
+                for(frame = 0; frame<DEFDACBLKSIZE; frame++)
+                    *fp++ = (t_sample)buf[chan + frame * channels];
```

Comment:
`float*buf = pulse_source->buffer;` can be const: `const float*buf = pulse_source->buffer;`

Please optimize the fragment as shown here:

```c
            int chan;
            const float*buf = pulse_source->buffer;
            if (channels == 1 && (sizeof(*fp) == sizeof(*buf)))
                memcpy(fp, buf, DEFDACBLKSIZE * sizeof(*fp));
            else {
                t_sample*lim;
                for(chan = 0; chan<channels; chan++) {
                    buf = pulse_source->buffer + chan;
                    lim = fp + DEFDACBLKSIZE;
                    for (; fp < lim; buf += channels) {
                        *fp++ = (t_sample)*buf;
                    }
                }
            }
```

or if you want to be "fancy" (but less readable), then you can use loop unrolling with duff's device:

```c
            int chan;
            const float*buf = pulse_source->buffer;
            if (channels == 1 && (sizeof(*fp) == sizeof(*buf)))
                memcpy(fp, buf, DEFDACBLKSIZE * sizeof(*fp));
            else {
                t_sample*lim;
                for(chan = 0; chan<channels; chan++) {
                    buf = pulse_source->buffer + chan;
                    lim = fp + DEFDACBLKSIZE;
                    // optimization: Duff's device
                    switch (DEFDACBLKSIZE % 4) {
                        case 0:
                            do {
                                *fp++ = (t_sample)*buf; buf += channels;
                        case 3:
                                *fp++ = (t_sample)*buf; buf += channels;
                        case 2:
                                *fp++ = (t_sample)*buf; buf += channels;
                        case 1:
                                *fp++ = (t_sample)*buf; buf += channels;
                            } while (fp < lim);
                    }
                }
            }
```